### PR TITLE
8335062: NMT: Make StackIndex non-opaque

### PR DESCRIPTION
**Pull request by jdksjolen**
Having SI being a opaque struct turned out to be a cumbersome idea, just make it an `int` to match GrowableArray's index type.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335062](https://bugs.openjdk.org/browse/JDK-8335062): NMT: Make StackIndex non-opaque (**Enhancement** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19884/head:pull/19884` \
`$ git checkout pull/19884`

Update a local copy of the PR: \
`$ git checkout pull/19884` \
`$ git pull https://git.openjdk.org/jdk.git pull/19884/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19884`

View PR using the GUI difftool: \
`$ git pr show -t 19884`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19884.diff">https://git.openjdk.org/jdk/pull/19884.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp [hunk 1]
```diff
@@ -41,19 +42,17 @@
...
 
-    bool is_invalid() {
-      return _stack_index == invalid;
-    }
-  };
+  constexpr static const StackIndex invalid = std::numeric_limits<StackIndex>::max() - 1;
+
+  static bool equals(const StackIndex a, const StackIndex b) {
+    return a == b;
+  }
```
Comment (jdksjolen) [inline src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp:51]: Can get rid of, but is also fine imho.

**Review by tstuefe — APPROVED** [review]
Looks good

File: src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp (hunk 1 above)
Comment (gerard-ziemski) [reply src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp:51]: RegionData uses it in:

```
    static bool equals(const RegionData& a, const RegionData& b) {
      return a.flag == b.flag &&
             NativeCallStackStorage::StackIndex::equals(a.stack_idx, b.stack_idx);
    }
```

so I'd say it's nice to have it encapsulated like this.

File: src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp [hunk 2]
```diff
@@ -41,19 +42,17 @@
...
-    static bool equals(const StackIndex& a, const StackIndex& b) {
-      return a._stack_index == b._stack_index;
-    }
+  using StackIndex = int;
 
-    bool is_invalid() {
-      return _stack_index == invalid;
-    }
-  };
+  constexpr static const StackIndex invalid = std::numeric_limits<StackIndex>::max() - 1;
```
Comment (gerard-ziemski) [inline src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp:47]: I'm trying to figure out whether this needs to be `public` and even though I see other classes using `StackIndex`, I don't see them actually needing to check the value.

I think it's fine to leave it `public`, though I would have started with `private` and made it `public` only if strictly needed.

Though I'd probably prefer if it was called `InvalidIndexValue` or something better than just `invalid`.

**Review by gerard-ziemski — APPROVED** [review]
Looks good, just a small nitpick if you care to address. Thanks!

File: src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp (hunk 2 above)
Comment (jdksjolen) [reply src/hotspot/share/nmt/nmtNativeCallStackStorage.hpp:48]: Yeah, we can just leave it private and force a PR which needs it to make it public. Let's keep the simple name for now, we'll see how readable it is when/if it becomes a public field.

**Comment by jdksjolen** [comment]
@tstuefe , @gerard-ziemski , ping for re-approval in accordance with the new rules!

**Review by gerard-ziemski — APPROVED** [review]
LGTM, thanks for fixing this!

**Comment by jdksjolen** [comment]
@tstuefe , could you please re-review this? Thank you.

**Review by tstuefe — APPROVED** [review]
Still looks good. Please make sure GHAs are either green or erorrs are unrelated (and platforms are tested).

**Comment by jdksjolen** [comment]
GHA passes, thank you.

/integrate

**Pull request closed without merging**